tools/wake-format/wake-format.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include <cstdio>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+#include "gopt/gopt-arg.h"
+#include "gopt/gopt.h"
+#include "parser/cst.h"
+#include "parser/syntax.h"
+#include "util/diagnostic.h"
+#include "util/file.h"
+
+#define VERSION "0.0.1"
+
+class TerminalReporter : public DiagnosticReporter {
+ public:
+  TerminalReporter() : errors(false), warnings(false) {}
+  bool errors;
+  bool warnings;
+
+ private:
+  std::string last;
+
+  void report(Diagnostic diagnostic) {
+    if (diagnostic.getSeverity() == S_ERROR) errors = true;
+    if (diagnostic.getSeverity() == S_WARNING) warnings = true;
+
+    if (last != diagnostic.getMessage()) {
+      last = diagnostic.getMessage();
+      std::cerr << diagnostic.getLocation() << ": ";
+      if (diagnostic.getSeverity() == S_WARNING) std::cerr << "(warning) ";
+      std::cerr << diagnostic.getMessage() << std::endl;
+    }
+  }
+};
+
+void print_help(const char *argv0) {
+  // clang-format off
+  std::cout << std::endl
+    << "Usage: " << argv0 << " [OPTIONS] [<file> ...]" << std::endl
+    << "  --dry-run  -n     Check if formatting needed, but don't apply it"      << std::endl
```

Comment:
love this option.

Reply:
The dry-run option can be later used by CI systems so we can prevent unformatted code from getting into codebases